internal/chain/reward.go

```diff
@@ -27,34 +27,65 @@ func ballotResultBeaconchain(
 	return availability.BallotResult(bc, header, shard.BeaconChainShardID)
 }
 
+var (
+	votingPowerCache singleflight.Group
+)
+
+func lookupVotingPower(
+	epoch, currentEpoch *big.Int, subComm *shard.Committee,
+) (*votepower.Roster, error) {
+	key := fmt.Sprintf("%s-%d", epoch.String(), subComm.ShardID)
+	results, err, _ := votingPowerCache.Do(
+		key, func() (interface{}, error) {
+			votingPower, err := votepower.Compute(subComm)
+			if err != nil {
+				return nil, err
+			}
+			return votingPower, nil
+		},
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if new(big.Int).Sub(currentEpoch, epoch).Cmp(common.Big3) == 1 {
```

Comment:
If the epoch is not looked up. even if it passes 3 epoch old, it will never be cleared from cache?

Reply:
Right but via crosslinks you’ll usually end up hitting this case. We could have a time based approach as well or instead as a fall back or replacement of logic